src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs

```diff
@@ -806,27 +831,60 @@ public override IReadOnlyList<Type> GetParameterTypes(string methodName)
+        if (activitySource is null)
         {
-            var requestContext = connectionContext.OriginalActivity?.Context;
+            return null;
+        }
 
-            return signalRActivitySource.ActivitySource.StartActivity($"{_fullHubName}/{methodName}", ActivityKind.Server, parentId: null,
-                // https://github.com/open-telemetry/semantic-conventions/blob/main/docs/rpc/rpc-spans.md#server-attributes
-                tags: [
-                    new("rpc.method", methodName),
-                    new("rpc.system", "signalr"),
-                    new("rpc.service", _fullHubName),
-                    // See https://github.com/dotnet/aspnetcore/blob/027c60168383421750f01e427e4f749d0684bc02/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs#L308
-                    // And https://github.com/dotnet/aspnetcore/issues/43786
-                    //new("server.address", ...),
-                    ],
-                links: requestContext.HasValue ? [new ActivityLink(requestContext.Value)] : null);
+        var loggingEnabled = logger.IsEnabled(LogLevel.Critical);
+        if (!activitySource.HasListeners() && !loggingEnabled)
+        {
+            return null;
         }
 
-        return null;
+        var name = $"{_fullHubName}/{methodName}";
+        IEnumerable<KeyValuePair<string, object?>> tags =
+        [
+            new("rpc.method", methodName),
+            new("rpc.system", "signalr"),
+            new("rpc.service", _fullHubName),
+            // See https://github.com/dotnet/aspnetcore/blob/027c60168383421750f01e427e4f749d0684bc02/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs#L308
+            // And https://github.com/dotnet/aspnetcore/issues/43786
+            //new("server.address", ...),
+        ];
+        IEnumerable<ActivityLink>? links = (linkedActivity is not null) ? [new ActivityLink(linkedActivity.Context)] : null;
+
+        var propagator = serviceProvider.GetService<DistributedContextPropagator>() ?? DistributedContextPropagator.Current;
+
+        Activity? activity;
+        if (headers != null)
+        {
+            activity = ActivityCreator.CreateFromRemote(
+                activitySource,
+                propagator,
+                headers,
+                static (object? carrier, string fieldName, out string? fieldValue, out IEnumerable<string>? fieldValues) =>
+                {
+                    fieldValues = default;
+                    var headers = (IDictionary<string, string>)carrier!;
+                    headers.TryGetValue(fieldName, out fieldValue);
+                },
+                name,
+                tags,
+                links,
+                loggingEnabled);
+        }
+        else
+        {
+            activity = activitySource.CreateActivity(name, ActivityKind.Server, parentId: null, tags: tags, links: links);
+        }
+
+        activity?.Start();
```

Comment:
>activity?.Start(); [](http://example.com/codeflow?start=8&length=18)

I see it now you start the activity here. ignore my other question regarding that.